pkg/objstore/oci/helper.go

```diff
@@ -0,0 +1,332 @@
+// Copyright (c) 2021, Oracle and/or its affiliates.
+// Licensed under the Apache License 2.0.
+
+package oci
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"github.com/oracle/oci-go-sdk/v46/common"
+	"github.com/oracle/oci-go-sdk/v46/objectstorage"
+	"github.com/oracle/oci-go-sdk/v46/objectstorage/transfer"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func DefaultTransport(config Config) *http.Transport {
+	return &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+			DualStack: true,
+		}).DialContext,
+
+		IdleConnTimeout:       time.Duration(config.HTTPConfig.IdleConnTimeout),
+		ResponseHeaderTimeout: time.Duration(config.HTTPConfig.ResponseHeaderTimeout),
+		TLSHandshakeTimeout:   time.Duration(config.HTTPConfig.TLSHandshakeTimeout),
+		ExpectContinueTimeout: time.Duration(config.HTTPConfig.ExpectContinueTimeout),
+		TLSClientConfig:       &tls.Config{InsecureSkipVerify: config.HTTPConfig.InsecureSkipVerify},
+		MaxIdleConns:          config.HTTPConfig.MaxIdleConns,
+		MaxIdleConnsPerHost:   config.HTTPConfig.MaxIdleConnsPerHost,
+		MaxConnsPerHost:       config.HTTPConfig.MaxConnsPerHost,
+		DisableCompression:    config.HTTPConfig.DisableCompression,
+	}
+}
+
+func getAbsolutePath(bkt Bucket, objectName string) *string {
+	absolutePath := bkt.objectBasePath + objectName
+	return &absolutePath
+}
+
+func getRelativePath(bkt Bucket, objectName string) string {
+	relativePath := strings.TrimPrefix(objectName, bkt.objectBasePath)
+	return relativePath
+}
+
+func getNamespace(client objectstorage.ObjectStorageClient, requestMetadata common.RequestMetadata) (namespace *string, err error) {
+	response, err := client.GetNamespace(
+		context.Background(),
```

Comment:
In OCI GO SDK, the ObjectStorage Client already has a nice build it timeout and retry mechanism. It will be kind of redundant if we do it here. It was suggested by the internal OCI SDK team.
See https://github.com/oracle/oci-go-sdk/blob/5d8aab248378a0fcf7cab02926b696e6c8d34811/objectstorage/objectstorage_client.go#L983.

Reply:
AFAICT under the hood it uses the same `http.Transport` that we pass to it in the constructor and it has no timeouts? :thinking: Could you please show me where the timeouts are set for this operation? I don't think there is. But, I think we should stick with this `context.Background` because this is just a one-time operation on boot.

Reply:
httpClient is created by calling `DefaultTransport(config Config)`
```
	httpClient := http.Client{
		Transport: DefaultTransport(config),
	}
```

The DefaultTransport() below has `IdleConnTimeout`, `ResponseHeaderTimeout`, `TLSHandshakeTimeout`, and `ExpectContinueTimeout`.
```
func DefaultTransport(config Config) *http.Transport {
	return &http.Transport{
		Proxy: http.ProxyFromEnvironment,
		DialContext: (&net.Dialer{
			Timeout:   30 * time.Second,
			KeepAlive: 30 * time.Second,
			DualStack: true,
		}).DialContext,

		IdleConnTimeout:       time.Duration(config.HTTPConfig.IdleConnTimeout),
		ResponseHeaderTimeout: time.Duration(config.HTTPConfig.ResponseHeaderTimeout),
		TLSHandshakeTimeout:   time.Duration(config.HTTPConfig.TLSHandshakeTimeout),
		ExpectContinueTimeout: time.Duration(config.HTTPConfig.ExpectContinueTimeout),
		TLSClientConfig:       &tls.Config{InsecureSkipVerify: config.HTTPConfig.InsecureSkipVerify},
		MaxIdleConns:          config.HTTPConfig.MaxIdleConns,
		MaxIdleConnsPerHost:   config.HTTPConfig.MaxIdleConnsPerHost,
		MaxConnsPerHost:       config.HTTPConfig.MaxConnsPerHost,
		DisableCompression:    config.HTTPConfig.DisableCompression,
	}
}
```

Is that what you are looking for?

Reply:
No, I am talking about the timeout of this operation. The client has no timeout and we pass `context.Background` here so in case of a very slow TCP read, for example, this will be stuck indefinitely, right? :thinking: Perhaps we could add a log message before calling this function?

Reply:
Still missing a log message here before this call, with the `debug` log level.